### PR TITLE
[query] Refactor NDArray constructor interfaces

### PR DESCRIPTION
**Pull request by tpoterba**
* construct doesn't take structs, takes IndexedSeq[Value[Long]]
* Implement the nested loops in one place
* Delete some methods from type interface, insert PCanonical casts

### CODE REVIEW

File: hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala [hunk 1]
```diff
@@ -32,41 +33,19 @@ object LinalgCodeUtils {
...
-        dimVar := 0L,
-        Code.whileLoop(dimVar < shape(dimIdx),
-          innerLoops,
-          dimVar := dimVar + 1L
-        )
-      )
+    val (addElem, finish) = dataType.constructFromFunctions(cb, region, dataLength, deepCopy = false)
+    val idx = cb.newLocal[Int]("nda_create_column_major_idx", 0)
+    SNDArray.forEachIndex(cb, shape, "nda_create_column_major") { case (cb, idxVars) =>
+      addElem(cb, idx, IEmitCode.present(cb, pndv.loadElement(idxVars, cb).asPCode))
```
Comment (johnc1231) [inline hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala:44]: You're getting assertion errors from here. I think it's not ok to `IEmitCode.present` multiple times inside of an `IEmitCode.map`, which is what you're indirectly doing here within the NDArrayEmitter

**Review by johnc1231 — CHANGES_REQUESTED** [review]

File: hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala:44]: debugged that yesterday -- it's actually that IEmitCode.present is totally busted. I pulled out the fix in #9962

Comment (tpoterba) [reply hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala:44]: (assigned to you)

Comment (tpoterba) [reply hail/src/main/scala/is/hail/linalg/LinalgCodeUtils.scala:44]: I'm now getting failures about mismatched shapes in BLockMatrixDot examples (and others), I must some incorrect shape calculations.

**Comment by johnc1231** [comment]
Also some import conflicts that have to be resolved

**Comment by johnc1231** [comment]
Addressed your initial problem and merge conflicts, but looks like there's more that's going wrong. Seems like it's mostly data comparison failures with reasonableish looking data, probably a striding issue.

**Comment by johnc1231** [comment]
The problem is your loops are backwards. `SType.foreach` goes in row major order, and `NDArrayEmitter` expects to emit column major stuff.

**Comment by johnc1231** [comment]
Ok, I switched the implementation of `SNDArray.forEach` to the column major style we'd been using in the NDArrayEmitter previously, and now I think everything should pass.

**Review by johnc1231 — APPROVED** [review]

**Comment by johnc1231** [comment]
Anything that uses LAPACK seems broken. My guess is that this means `emitNDArrayColumnMajorStrides` is busted.

**Comment by johnc1231** [comment]
Definitely something funny happening here:

```
def emitNDArrayColumnMajorStrides(ir: IR): IEmitCode = {
      emitI(ir).map(cb){case pNDCode: PNDArrayCode =>
        val pNDValue = pNDCode.memoize(cb, "ndarray_column_major_check")
        val isColumnMajor = LinalgCodeUtils.checkColumnMajor(pNDValue, cb)
        val pAnswer = cb.emb.newPField("ndarray_output_column_major", pNDValue.pt)
        cb.ifx(false, {cb.assign(pAnswer, pNDValue)},
        {
          cb.println("Shoud be making a copy")
          cb.assign(pAnswer, LinalgCodeUtils.createColumnMajorCode(pNDValue, cb, region.code))
        })
        val ndarrayColMajor = pAnswer.asNDArray.memoize(cb, "temp")
        cb.println("col major output")
        cb.println(pNDValue.strides(cb).map(_.toS.concat(" ")):_*)
        cb.println("col major output over")
        pAnswer.get
      }
    }
```

The prints are coming out with row major striding, even though prints internal to `LinalgCodeUtils` look correct.

**Comment by johnc1231** [comment]
I believe everything works now, the problem was that `dataAddress` function was returning address of first element of data array, but we were acting like it was returning a pointer to the length of the array. Changed the name and the assumptions, and now it all works.

**Comment by tpoterba** [comment]
thank you John!

**Pull request merged**